api/src/main/java/org/apache/iceberg/types/Types.java

```diff
@@ -447,15 +438,77 @@ public static NestedField of(int id, boolean isOptional, String name, Type type,
       return new NestedField(isOptional, id, name, type, doc, null, null);
     }
 
-    public static NestedField of(
-        int id,
-        boolean isOptional,
-        String name,
-        Type type,
-        String doc,
-        Object initialDefault,
-        Object writeDefault) {
-      return new NestedField(isOptional, id, name, type, doc, initialDefault, writeDefault);
+    public static Builder from(NestedField field) {
+      return new Builder(field);
+    }
+
+    public static Builder required(String name) {
+      return new Builder(false, name);
+    }
+
+    public static Builder optional(String name) {
+      return new Builder(true, name);
+    }
+
+    public static class Builder {
+      private final boolean isOptional;
+      private final String name;
+      private Integer id = null;
+      private Type type = null;
+      private String doc = null;
+      private Object initialDefault = null;
+      private Object writeDefault = null;
+
+      private Builder(boolean isOptional, String name) {
+        this.isOptional = isOptional;
+        this.name = name;
+      }
+
+      private Builder(NestedField toCopy) {
+        this.isOptional = toCopy.isOptional;
+        this.name = toCopy.name;
+        this.id = toCopy.id;
+        this.type = toCopy.type;
+        this.doc = toCopy.doc;
+        this.initialDefault = toCopy.initialDefault;
+        this.writeDefault = toCopy.writeDefault;
+      }
+
+      public Builder withId(int id) {
+        this.id = id;
+        return this;
```

Comment:
There is already `withFieldId()`. How about deprecating/removing it before it spreads?

Reply:
It's part of the public API. We can deprecate it, but we can't remove it until v2.